### PR TITLE
Nominal `unique type` brands

### LINKED ISSUE #202: Support some non-structural (nominal) type matching
Proposal: support non-structural typing (e.g. new user-defined base-types, or some form of basic nominal typing). This allows programmer to have more refined types supporting frequently used idioms such as: 

1) Indexes that come from different tables. Because all indexes are strings (or numbers), it's easy to use the an index variable (intended for one table) with another index variable intended for a different table. Because indexes are the same type, no error is given. If we have abstract index classes this would be fixed.

2) Certain classes of functions (e.g. callbacks) can be important to be distinguished even though they have the same type. e.g. "() => void" often captures a side-effect producing function. Sometimes you want to control which ones are put into an event handler. Currently there's no way to type-check them. 

3) Consider having 2 different interfaces that have different optional parameters but the same required one. In typescript you will not get a compiler error when you provide one but need the other. Sometimes this is ok, but very often this is very not ok and you would love to have a compiler error rather than be confused at run-time.

Proposal (with all type-Error-lines removed!): 

```
// Define FooTable and FooIndex
nominal FooIndex = string;  // Proposed new kind of nominal declaration.
interface FooTable {
  [i: FooIndex]: { foo: number };
}
let s1: FooIndex;
let t1: FooTable;

// Define BarTable and BarIndex
nominal BarIndex = string; // Proposed new kind of nominal declaration.
interface BarTable {
  [i: BarIndex]: { bar: string };
}
let s2: BarIndex;
let t2: BarTable;

// For assignment from base-types and basic structures: no type-overloading is needed.
s1 = 'foo1';
t1 = {};
t1[s1] = { foo: 1 };

s2 = 'bar1';
t2 = { 'bar1': { bar: 'barbar' }};

console.log(s2 = s1); // Proposed to be type error.
console.log(s2 == s1); // Proposed to be type error.
console.log(s2 === s1); // Proposed to be type error.

t1[s2].foo = 100; // Gives a runtime error. Proposed to be type error.
t1[s1].foo = 100;

function BadFooTest(t: FooTable) {
  if (s2 in t) {  // Proposed to be type error.
    console.log('cool');
    console.log(t[s2].foo); // Proposed to be type error.
  }
}

function GoodBarTest(t: BarTable) {
  if (s2 in t) {
    console.log('cool');
    console.log(t[s2].bar);
  }
}

BadFooTest(t1); // Gives runtime error;
BadFooTest(t2); // No runtime error, Proposed to be type error.
GoodBarTest(t1); // Gives runtime error; Proposed to be type error.
GoodBarTest(t2);
```

### LINKED ISSUE #4895: Tag types
Problem:

- There is no straight way to encode a predicate (some checking) about a value in its type.

Details:

There are situation when a value has to pass some sort of check/validation prior to being used. For example: a min/max functions can only operate on a non-empty array so there must be a check if a given array has any elements. If we pass a plain array that might as well be empty, then we need to account for such case inside the min/max functions, by doing one of the following:
- crashing
- returning undefined
- returning a given default value
This way the calling side has to deal with the consequences of min/max being called yet not being able to deliver.

```typescript
function min<a>(values: a[], isGreater: (one: a, another: a) => boolean) : a {
   if (values.length < 1) { throw Error('Array is empty.'); }
   // ...
}
var found;
try {
   found = min([]);
} catch (e) {
   found = undefined;
}
```

A better idea is to leverage the type system to rule out a possibility of the min function being called with an empty array. In order to do so we might consider so called tag types. 

A **tag type** is a qualifier type that indicates some predicate about a value it is associated with holds true.

```typescript

const enum AsNonEmpty {} // a tag type that encodes that a check for being non-empty was passed

function min<a>(values: a[] & AsNonEmpty) : a {
   // only values tagged with AsNonEmpty can be passed to this function
   // the type system is responsible for enforcing this constraint
   // a guarantee by the type system that only non-empty array will be given makes
   // implementation of this function simpler since only one main case needs be considered
   // leaving handling the empty-case situation outside of this function
}
```

it's up to the developer in what circumstances an array gets its AsNonEmpty tag, which can be something like:

```typescript

interface GuaranteedByServer {
    values: number[] & AsNonEmpty;
}

```

```typescript

interface 

```

As was shown in the current version (1.6) an empty const enum type can be used as a marker type (AsNonEmpty in the above example), however it has some limitations:

- enum as assignable by numbers
- enum cannot hold a type parameter
- enum cannot have members

A few more examples of what tag type can encode:

- `string & AsTrimmed & AsLowerCased & AsAtLeast3CharLong`
- `number & AsNonNegative & AsEven`
- `date & AsInWinter & AsFirstDayOfMonth`

Custom types can also be augmented with tags. This is especially useful when the types are defined outside of the project and developers can't alter them.

- `User & AsHavingClearance`

**ALSO NOTE**: In a way tag types are similar to boolean properties (flags), BUT they get type-erased and carry no rutime overhead whatsoever being a good example of a zero-cost abstractions.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #202
Fixes #4895

We've talked about this on and off for the last three years, and it was a major reason we chose to use `unique symbol` for the individual-symbol-type, since we wanted to reuse the operator for a nominal tag later. What this PR allows:
```ts
type NormalizedPath = unique string;
type AbsolutePath = unique string;
type NormalizedAbsolutePath = NormalizedPath & AbsolutePath;

declare function isNormalizedPath(x: string): x is NormalizedPath;
declare function isAbsolutePath(x: string): x is AbsolutePath;
declare function consumeNormalizedAbsolutePath(x: NormalizedAbsolutePath): void;


const p = "/a/b/c";
if (isNormalizedPath(p)) {
    if (isAbsolutePath(p)) {
        consumeNormalizedAbsolutePath(p);
    }
}
```

`unique T` (where `T` is any type) is allowed in any position a type is allowed, and nominally tags `T` with a marker that makes only `T`'s that have come from that location be assignable to the resulting type.

This is done by adding a new `NominalBrand` type flag, which is a type with no structure which is unique to each symbol it is manufactured from. This is then mixed into the argument type to `unique type` via intersection, which is what produces all useful relationships. (The brand can have an alias if it is directly constructed via `type MyBrand = unique unknown`)

This does so much with so little - this reduces the jankiness written into types to enable nominalness with `unique symbol`s or `enum`s, while adding _zero_ new assignability rules.

So, why bring this up now? I was thinking about how "brands" work today, with something like `type MyBrand<T> = T & {[myuniquesym]: void}` where `T` could then become a literal type like `"a"`. We've wanted, for awhile, to be able to more eagerly reduce an intersection of an object literal and a primitive to `never` (to make subtype reduction and intersection reduction produce less jank and recognize more types as mutually exclusive), but these "brand" patterns keep stopping us. (Heck, we use em internally.) Well, if we ever want to change object types to actually mean _object_, then we're going to need to provide an alternative for the brand pattern, and ideally that alternative needs to be available for _awhile_. So looking on the horizon to breaks we could take into 4.0 in 9 months, this simplification of branding would be up there, provided we've had the migration path available for awhile. So I'm trying to get the conversation started on this before we're too close to that deadline to plan something like that. Plus #202 is up there on our list of all-time most requested issues, and while we've always been open to it, we've never put forward a proposal of our own - well, here one is.

On `unique symbol`
---------------------
`unique symbol`'s current behavior takes priority for syntactically exactly `unique symbol`, however if a nominal subclass of symbols is actually desired, one can write `unique (symbol)` to get a nominally branded `symbol` type instead (or `symbol & unique unknown` - it's exactly the same thing). The way a `unique symbol` behaves is like a symbol that is locked to a specific declaration, and has special abilities when it comes to narrowing and control flow because of that. Nominally branded types are more flexible in their usage but do not have the strong control-flow linkage with a host statement that `unique symbol`s do (in fact, they don't necessarily assume a value exists at all), so there is very much reason for them to coexist. They're _similar enough_, that I'm pretty comfortable sharing syntax between the two.

Alternative considerations
----------------------------
While I've used the `unique` type operator here, like we've oft spoken of, on implementation, it's become plain to me that _I don't need to specify an argument to `unique`_. We could just expose `unique` as a unique type factory on it's own, and dispense with the indirection. The "uniqueness" we apply internally isn't actually tied to the input type argument through anything more than an intersection type, so simply shortening `unique unknown` to `unique` and reserving the argument form for just `unique symbol`s may be preferable. All the same patterns would be possible, one would just need to write `string & unique` instead of `unique string`, thus dispensing with the sugar. It depends on the perceived complexity, I think. However, despite being _exactly the same_, `string & unique` is somehow uglier and harder to look at than `unique string`, which is why I've kept it around for now. It's probably worth discussing, though.

What this draft would still need to be completed:
* [ ] New error messages for any errors involving brands (`One or more unique brands is missing from type A` with related information pointing at the brand location, rather than the current error involving `unique unknown`)
* [ ] More tests exercising indexed access and exploring how indexed accesses on branded types are constructed (specifically, what should be done if you (attempt to) index nothing but a union of `unique unknown` brands)
* [ ] Tests for `unique (symbol)` declaration emit (to ensure it's not rewritten as `unique symbol`)
* [ ] Discussion on if `keyof <unique brand type>` should be `never` (as it is now), since the brand is top-ish (and contains no structure information itself), or if it should be preserved as an abstract `keyof <unique brand>` type, so that brand can apply keys-of-branded-types constraints in constraint positions
* [ ] 🚲 🏠

### CODE REVIEW

**Comment by fatcerberus** [comment]
> While I've used the unique type operator here, like we've oft spoken of, on implementation, it's become plain to me that I don't need to specify an argument to unique.

I don't understand this part.  If these are meant to replace branded types, then:

```ts
type UString = unique string;
type BString = string & { __brand: true };

declare let ustr: UString;
declare let bstr: BString;
let str1: string = bstr;  // ok because branded string is a subtype of string (this is generally desirable).
let str2: string = ustr;  // could be ok because unique string is also a string.
let num: number = ustr;  // never ok, should be error because unique string is NOT a number!
```

But if we just had `unique` without the argument, there would be no way to make this distinction.  `unique` would just end up being a nominal `unknown` (effectively an opaque Skolem constant) which doesn't sound that useful to me?

**Comment by weswigham** [comment]
> But if we just had unique without the argument, there would be no way to make this distinction.

It's useful because you can then intersect it with something. Which is all `unique string` is doing under the hood right now - making that intersection for you.

**Comment by fatcerberus** [comment]
Yeah, I need to read more closely.  I just noticed the `string & unique` bit before you posted.  But I don't like this because `unique` by itself *isn't a type*.  It would be a special marker that doesn't work by itself as a type (or at least, doesn't make sense as one - what would it mean to take an argument of form `arg: unique`, e.g.), which I find very weird to then use in a position that nominally (no pun intended!) accepts only type operands.  I guess there's precedent with `ThisType<T>`, but... I don't know, it rubs me the wrong way.

**Comment by weswigham** [comment]
`unique` is essentially ust shorthand for `unique unknown` under the current model, and it does function fully as a standalone type.

**Comment by cevek** [comment]
Is this proposal allows to assign literals to variable/param which has branded type?
```ts
type UserId = unique string
function foo(param: UserId) {}

foo("foo") // ok
var x: UserId = "foo"; // ok

const s = "str";
var y: UserId = s; // not ok
foo(s) // not ok
```

**Comment by goodmind** [comment]
How would you make nominal classes with this?

**Comment by weswigham** [comment]
Not easily. You'd need to mix a distinct nominal brand into every class declaration, like via a property you don't use of type `unique unknown` or similar, and then ensure that your inheritance trees always override such a field (so a parent and child don't appear nominally identical).

I'd avoid it, if possible, tbh. Nominal classes sound like a pain :P

**Comment by weswigham** [comment]
> Is this proposal allows to assign literals to variable/param which has branded type?

Branded types can only be "made" either via cast or typeguard, as I said in the OP, so no. This is because the typesystem doesn't know what invariants a given brand is meant to maintain, and can't implicitly know if some literal satisfies them.

**Comment by jack-williams** [comment]
Would something like the following ever be meaningful?
```ts
interface Parent extends unique { }
interface ChildA extends (Parent & unique) { }
interface ChildB extends (Parent & unique) { }
```

**Comment by AnyhowStep** [comment]
This is probably obvious to everyone but `unique T` shouldn't replace branding.
There are scenarios where branding is the only viable solution (at the moment).

For example,
```ts
//Can be replaced with `unique number`
type LessThan256 = number & { __rangeLt : 256 }
//Cannot be replaced with `unique number`
type LessThan<N extends number> = number & { __rangeLt : N }
```

More complicated example here,
https://github.com/microsoft/TypeScript/issues/15480#issuecomment-513105767

-----

There are a few reasons why I'm generally against the idea of `unique T`.
(And nominal typing, and `instanceof`, and using symbols)

### Cross-library interop

Library A may have `type Radian = unique number`.
Library B may have `type Radian = unique number`.

Both types will be considered different, even though they have the same name and declaration.

If libraries start using this `unique T` all over the place, you'll start needing unsafe casts (`as libA.Radian`) more often. This means you can accidentally write `myDegree as libA.Radian` and cast incorrectly. Whoops!

So, one starts thinking that a no-op casting function would be safer,
```ts
function libARadianToLibBRadian (rad : libA.Radian) : libB.Radian {
  return rad as libB.Radian;
}
```

This is safer because you won't accidentally convert `myDegree:libA.Degree` to `libB.Radian`

But if you have `N` libraries with their own `Radian` type, you may end up needing up to `N^2` functions to convert between each of the `Radian` types from each library.

-----

### Cross-version interop

It's happened to me a bunch where I've had **the same** package, but at different versions, within a single project.

So,

v1.0.0's `type Radian = unique number` would be considered different from,
v1.1.0's `type Radian = unique number`

Now you need a casting function... Even though it's the same package.

-----

With brands, if two libraries use the same brands, even if they're different types, they'll still be assignable to each other. (As long as they don't use `unique symbol`)


Library A may have `type Rad = number & { __isRadian : void }`.
Library B may have `type Radian = number & { __isRadian : void }`.

Even though they're different types, they're assignable to each other. No casting needed.
v1.0.0 and v1.1.0 of `type Rad` and `type Radian` will work with each other fine.

**Comment by fatcerberus** [comment]
Re: nominal classes - classes are *already* nominal if they contain any private members.  Just throwing that out there. :smiley:

**Comment by be5invis** [comment]
So we can *finally* have things like this?

```typescript
// Low-end refinement type :)
type NonEmptyArray<A> = unique ReadonlyArray<A>
function isNonEmpty(a: ReadonlyArray<A>): a is NonEmptyArray<A> {
    return a.length > 0
}
```

**Comment by AnyhowStep** [comment]
@fatcerberus It's also why I avoid classes entirely and avoid private members if I do have them =P

**Comment by AnyhowStep** [comment]
Hmm...
```ts
const normalizedPathBrand = Symbol();
const absolutePathBrand = Symbol();

type NormalizedPath = string & typeof normalizedPathBrand;
type AbsolutePath = string & typeof absolutePathBrand;
type NormalizedAbsolutePath = NormalizedPath & AbsolutePath;

declare function isNormalizedPath(x: string): x is NormalizedPath;
declare function isAbsolutePath(x: string): x is AbsolutePath;
declare function consumeNormalizedAbsolutePath(x: NormalizedAbsolutePath): void;


const p = "/a/b/c";
consumeNormalizedAbsolutePath(p); //Error
if (isNormalizedPath(p)) {
    consumeNormalizedAbsolutePath(p); //Error
    if (isAbsolutePath(p)) {
        consumeNormalizedAbsolutePath(p); //OK
    }
}
```
[Playground](https://typescript-play.js.org/#code/MYewdgzgLgBGICcC2BDANgSwF4FMAmACilABYBCCKYeMAvDAMoCeSARiGgBQCUA3AFChIsFKwgcArlBxFSFKjXrM2HHgP5QmABxwwAcolSZchYiToxoCDGADmMAGQxNOkADM4h9NnyzylagEXXQBBMUlpPwsrG3snYPcYUXE0KRkzeUCNbV0DZG8TMJS0qPo8ox9TUkcYIoj00nU8HGA0FARdNwkwYCgMcBgMCHKC3zNOAA8ALksoaztuGYnBiH0vYzHG-mbW9s7u3v6wFbrUyPHp2fnbRZhlodrws4aSAR22jpgunr6BoQgJEgcCMNnhTiULjMQZVwedSLcAG4gDB4dSCcDQGBaCwAIgA9Cg8aw8cAcQJ-oDgesYU8IaROFo+DA8XiAKIIBCIfgYDycIbQkx+BncbgwADe-BgUpgFKBAvwsJewt4zLZHK50sGvKGiqFjNFEs1mtlVPyoN140ZKpZAHkANKS6UAX34TqAA)

**Comment by jack-williams** [comment]
If cross version compatibility really is an issue then an alternate solution would be to make naming explicit:

Let the type `unknown "name"` denote the set of all values with label _name_. This would make the intersection type approach mandatory so a nominal path must now be written:
```ts
type NormalizedPathOne = string & unique "NormalizedPath";
```
where an unlabelled `unique` denotes a generative type as defined in the OP.
```ts
type NormalizedPathTwo = string & unique // some label that we don't care about that is auto-generated.
```
so while two declarations of `NormalizedPathTwo` produce distinct types, two declarations of `NormalizedPathOne` produce identical types.

FWIW I have no real preference---for me the big win of this feature is being able reduce empty intersections more aggressively.

> Discussion on if keyof unique brand type ....

IMO, for all brand oblivious operations a `unique` type should be equivalent to `unknown` (which is what is proposed).

**Comment by resynth1943** [comment]
I've implemented Opaque types like so:
```ts
type Opaque<V> = V & { readonly __opq__: unique symbol };

type AccountNumber = Opaque<number>;
type AccountBalance = Opaque<number>;

function createAccountNumber (): AccountNumber {
    return 2 as AccountNumber;
}

function getMoneyForAccount (accountNumber: AccountNumber): AccountBalance {
    return 4 as AccountBalance;
}

getMoneyForAccount(100); // -> error
```

**Comment by resynth1943** [comment]
@AnyhowStep I know, but that's how I'm currently creating opaque types. I hope this Pull Request will incorporate this into the language, and make it even better than my implementation.

**Comment by mohsen1** [comment]
Nominal types are pretty useful. The example I often use is the APIs that take latitude/longitude and bugs that are result of mixing up latitude with longitude which are both numbers. By making those unique types we can avoid that class of bugs. 

However, unique types can cause so much pain when you have to keep importing those types to simply use an API. So I'm hoping that at least primitive types are assignable to unique primitives where I can still call my functions like this:

```ts
// lib.ts
export type Lat = unique number;
export type Lng = unique number;
export function distance(lat: Lat, lng: Lng): number;

// usage.ts
import { distance } from 'lib.ts';
distance(1234, 5678); // no need to asset types
```

As @AnyhowStep mentioned cross-lib and cross-version conflicting unique types can also be a source of pain. Can we limit uniqueness scope somehow? Would that be a viable solution?

**Comment by weswigham** [comment]
https://github.com/microsoft/TypeScript/pull/33290 is now open as well, so we can have a real conversation on what the non-nominal explicit tag would look like, and if we'd prefer it.

**Comment by weswigham** [comment]
BTW, this is now a dueling features type situation (though I've authored both) - we won't accept both a #33290 style brand and this PR's style brand, only one of the two (and we're leaning towards #33290 on initial discussion). We'll get to debating it within the team hopefully during our next design meeting (next friday), but y'all should express ideas, preferences, and reasoning therefore within both PRs.

**Comment by xiaoxiangmoe** [comment]
@typescript-bot pack this

**Comment by xiaoxiangmoe** [comment]
@weswigham Only the authors of pr can `@typescript-bot`?

**Comment by weswigham** [comment]
Only repo contributors have permission to request that @typescript-bot pack this

**Comment by xiaoxiangmoe** [comment]
How should I write a unique type constructor?
```ts
type integer = unique number
function isInteger(a: number) { return a === (a | 0) }

function Interger (a:number): integer{
    if(!isInteger(a)) throw new Error("not integer");
    return a // Error: Type 'number' is not assignable to type 'integer'. \n Type 'number' is not assignable to type 'unique unknown'.
}
```

**Comment by raveclassic** [comment]
@xiaoxiangmoe I think you should add a type guard: 
```typescript
function isInteger(a: number): a is integer { return return a === (a | 0) }
```

**Comment by ProdigySim** [comment]
Borrowing from [an example from the other PR](https://github.com/microsoft/TypeScript/pull/33290#issuecomment-529114949)...

Let's say I want to have a unique `UserId` type that is also tagged with details about its composition. What would be the correct way to combine these types into a final unique version?

```ts
type NonEmptyString = unique string;
type Uuid = unique string;

// Oops, these two are probably assignable to each other!
type GroupId = NonEmptyString & Uuid;
type UserId = NonEmptyString & Uuid;
```

I downloaded the tgz build and tried these out, which both appear to "work" on a surface level...

```ts
type GroupId = unique (NonEmptyString & Uuid);
type UserId = unique (NonEmptyString & Uuid);


type GroupId = unique string & NonEmptyString & Uuid;
type UserId = unique string & NonEmptyString & Uuid;
```

**Comment by weswigham** [comment]
@ProdigySim yep, the way you'd accomplish that is with extra `unique` tags for each `unique` thing you want to track.

**Comment by joshburgess** [comment]
How would one go about using phantom type params to get unique nominal types with this PR? Is there a way?

For example:

```typescript
type User = ...
type Organization = ...

Key<A> = unique string

type UserId = Key<User>
type OrganizationId = Key<Organization>
```

The goal of the above would be to make sure `UserId` and `OrganizationId` aren't interchangeable, but the generic type param does not influence the structure of the runtime type at all. Would the above just work? Or would you need to do something like:

```typescript
Key<A> = unique string & A
```

or something else?

**Comment by weswigham** [comment]
> How would one go about using phantom type params to get unique nominal types with this PR? Is there a way?

As is, nope. A `unique` tag is unique per declaration site, and every instantiation gets the _same_ unique brand, irrespective of type parameters. You'd need to do something like `type Key<A> = unique string & A`, except you _probably_ want to stash `A` somewhere it's value can't be observed (like in a property). The phantom type use-case of the structural brands we have today isn't handled well by this proposal.

**Comment by joshburgess** [comment]
@weswigham I just saw your other PR using `tag` which does support phantom types. That looks interesting.

Just out of curiosity, would it be possible to add a keyword solely for this purpose to be used in tandem with this implementation? 

Maybe, something like:

```typescript
type Key<A> = unique string & phantom A
```

or 

```typescript
type Key<A> = unique string & tagged A
```

or something like that, where `phantom` or `tagged` essentially behave similarly to `unique`, uniquely modifying the nominal type, but without any false/ghost properties added to the structural type.

Alternatively, maybe using `unique` as a prefix to a type alias would implicitly do something just like the above? like:

```typescript
// the leading `unique` automatically affects the nominal type under the hood
unique type Key<A> = unique string
```

**Comment by resynth1943** [comment]
I'm not familiar with all this rhetoric, bit isn't having phantom types the *goal*, as they're actually unique types? Theyre without any `__mysterious_props__` (and weird compiler errors).

**Comment by mheiber** [comment]
Will `unique`

Currently, the best I can think of for declaration file emit for:

```ts
// example.ts
export class A {
     #foo: number;
}
```

is:

```ts
// example.d.ts
export declare class A {
    #foo;
}
```

But this violates the spirit of hash private fields: users shouldn't have to scroll past like 50 implementation details in order to see the stuff that matters to them.


If we can emit `unique` that would preserve the nominality and hide the implementation better:

```ts
// example.d.ts
export declare unique class A {}
```

**Comment by treybrisbane** [comment]
@weswigham Thanks for doing this work and creating this proposal! The `unique` approach is a much better fit for my requirements than the `tag` proposal. :D

Just out of curiosity, have you and/or the team considered having `unique` be a "declaration transformer" rather than a "type transformer"?

E.g.

```ts
unique type Foo = string;
// Rather than `type Foo = unique string;`

unique interface Bar {
  field: string;
}

unique class Baz {
  constructor(
    public field: string,
  ) {}
}
```

Assuming that your underlying brand intersection implementation can stay the same, I can't help but notice that a "declaration transformer" approach seems to have several advantages over the "type transformer" approach:

- It opens up the possibility of nominal interfaces and classes (which several people have already asked about)
- It's syntactically closer to the prior art of Flow's opaque types (`opaque type Foo = string;`, IIRC)
- It more clearly distinguishes between nominal types and `unique symbol`s (which, as you've said, are meaningfully different, and I subjectively don't believe that using parens for that case is intuitive)

If you guys have looked at this approach and decided against it, could you elaborate as to why?

**Comment by joshburgess** [comment]
@treybrisbane That is also the syntax I would most like, I think... especially if it automatically made any type params present affect the nominal type, even if they were phantom type params.

`unique type Key<A> = string`

In the above, every unique `A`, regardless of whether or not it gets used in the right side of the type definition, would result in a unique nominal type.

I don't know how doable this is, and, perhaps, it would need to be implemented as a structural type hack under the hood, but it would be nice to be able to support phantom types like this without requiring the user to think about structural types (like the `tag` in the other PR, or similar).

This might be hard/not possible. Not sure. Just thinking out loud.

**Comment by xiaoxiangmoe** [comment]
Is there any progress on this feature?

**Comment by ExE-Boss** [comment]
Will&nbsp;this&nbsp;support <code>unique&nbsp;interface</code>&nbsp;types as&nbsp;well?

**Comment by michaeljota** [comment]
I like this proposal more than the other one. This is more explicit and less verbose. I understand the concerns raised here, but I think they should be considered features, and this should be used carefully by the community. I understand this is not always the case, and people could start using and misusing this for everything, but this is as for all the features that Typescript introduces. They are not always used as intended.

One particularly seems to be a good example of why this proposal is good for me. 

## Cross-library interop
Library A may have `type Radian = unique number`.
Library B may have `type Radian = unique number`.

And they _are_ different. Libary A can update is definition tomorrow to `type Radian = bigint`, and that would be ok. I understand that _if I have two nominal types, that both are numbers, seems wrong that they can't assign each other_, but is that what this is about? Having someway to declare a _unique type_?

If I can declare a _unique type_ that can be assigned to a different _unique type_ because they have the same shape, are they unique then?

**Comment by AnyhowStep** [comment]
This PR leads to the Tower of Babel problem more easily, because of cross-library interop.

**Comment by Shou** [comment]
@AnyhowStep 

> But if you have N libraries with their own Radian type, you may end up needing up to N^2 functions to convert between each of the Radian types from each library.

You don't have to create `N^2` converting functions, just one. In Haskell, Conor McBride's `newtype` library uses `over` and the like to allow conversion between newtypes. Now, we don't have typeclasses, but we could still do something similar here:

```
const coerce: <A, B>(a: unique A) => IfEquals(A, B, unique B, never)
  = (a) => a as any
```

Where the `IfEquals` function is taken from [here](https://stackoverflow.com/questions/53807517/how-to-test-if-two-types-are-exactly-the-same). We could also use `as ReturnType<typeof coerce>` instead of `as any` but it makes no difference here. Now you can convert between any `unique`s that share the same underlying type:

```
coerce<libA.Radian, libB.Radian>(foo)
```

**Comment by AnyhowStep** [comment]
`coerce<,>()`, `foo as libB.Radian`, and the N^2 conversion functions are all unsafe, of course.

However, the N^2 explicit functions is "safer" because **more effort** had to go into creating them.

Having a `coerce<,>()` function is much closer to the `as` operator, and is more "unsafe" because it lowers the barrier to casting types.

That's my personal opinion, anyway.

**Comment by shicks** [comment]
To what extent does https://github.com/microsoft/TypeScript/issues/31894 satisfy the same requirements as this proposal?  I know @DanielRosenwasser mentioned in https://github.com/microsoft/TypeScript/issues/38510 that while nominal brands would be nice, but that it's not clear whether it's counter to the goals of 4.0 at the moment.

That said, it seems to me that one could easily enough write
```
placeholder type NormalizedPathBrand;
placeholder type AbsolutePathBrand;
type NormalizedPath = string & NormalizedPathBrand;
type AbsolutePath = string & AbsolutePathBrand;
```
where the brands are module-local and never intended to be implemented by any concrete type.  Branded types already need explicit casts to instantiate them, so it's easy enough to make them.  Given that that proposal seems to have more traction, does it make this one obsolete?

**Comment by isiahmeadows** [comment]
Is there a status update on this PR?

**Comment by isiahmeadows** [comment]
@weswigham Couple questions:

1. Could you elaborate further on how `unique symbol` and `unique (symbol)` differ with respect to control flow-sensitive typing? I would've thought neither would have special behavior on that front.
2. Do you have any status update for this? It's been over a year with pretty much radio silence.

**Comment by weswigham** [comment]
>Could you elaborate further on how unique symbol and unique (symbol) differ with respect to control flow-sensitive typing? I would've thought neither would have special behavior on that front.

`unique symbol` is a _single_ runtime value. Specifically, the result of an invocation of `Symbol()` at some location. Because of this, it works as a well-known property name, and acts as a literal. Each place `unique symbol` is written is a distinct, single symbol. `unique (symbol)` is "a unique subset of all symbols" and represents an abstract set of potentially _multiple_ symbols. This behaves differently - notably, it isn't usable as well-known property name.

> Do you have any status update for this? It's been over a year with pretty much radio silence.

Oh, wow, it's been that long. Uh. Last time I presented it to the team, the response was somewhat lukewarm - there's no real excitement or drive for it right now. So I guess what we're looking to see is overwhelming demand?

**Comment by kachkaev** [comment]
> So I guess what we're looking to see is overwhelming demand?

Not sure how the demand is properly measured, just saying that this PR is in [top 5 upvoted open PRs](https://github.com/microsoft/TypeScript/pulls?q=is%3Apr+is%3Aopen+sort%3Areactions-%2B1-desc) at them moment 😁

**Comment by isiahmeadows** [comment]
@weswigham Thanks for the quick explanation! I see why both are necessary, now. Also, I definitely recommend looking into @kachkaev's comment above.

**Comment by alex-weej** [comment]
Just used branding to very quickly identify the inconsistencies in a code base that was using string for 3 different logical dynamically valued types. Consider this my +1...

**Review by mostasim — APPROVED** [review]

**Comment by ProdigySim** [comment]
Just a bit of data: The last codebase I was in changed from using `unique symbol` to string-based tag types due to their simplicity in creating compatible types across repos.

We wanted to move to an code generation system for our web API types. We had backend code that used nominal types ("tinytypes") in Scala which was the source of our branding.

Using unique-symbol based types had a couple of drawbacks here:
1. We would need to have every nominal type exposed on the API to be predefined somewhere in TypeScript
2. If the backend & frontend nominal type names mismatched at all, we would need to write custom mapping.
3. One potential solution would be to have a shared library with all our "global" nominal types. But, this opens the possibility for "Dreaded Diamond" dependency patterns on these types, which would lead to incompatible nominal types (different "unique symbols" from different versions of the library)

We ended up converting all of our nominal types to string-based tag types overnight to support this project. They solved all of these problems:
1. API code gen can just emit a generic `Opaque<T, 'tag'>` and not worry about references/imports
2. If backend and frontend nomenclature mismatch, we can negotiate on the value of the `'tag'` string, without any user-facing implications.
3. Don't need a shared library in this solution. If we did opt for one, the dreaded diamond would not be an issue unless the type tags actually changed--which would probably mark an important compatibility change.

tl;dr string-based tag types seem a little more flexible with implementation details and are still good enough. They still have the hard problem of naming things, but they are fixable when there are name conflicts. We can also put out recommendations for tag patterns in shared libraries.

**Comment by sandersn** [comment]
This experiment is pretty old, so I'm going to close it to reduce the number of open PRs.

**Comment by negue** [comment]
Will there be any type of these `unique`/`opaque` types or how ever these are called be added into typescript itself?

I needed something like that in a project where I had to juggle like a couple of simple "number" IDs and here something like unique number type per ID would've helped to save some time debugging :D

**Comment by shicks** [comment]
Is there a particular reason this experiment stalled out?

**Comment by evelant** [comment]
@sandersn IMO this is a very valuable feature. Nominal types can greatly increase type safety and I would imagine they could also greatly improve compiler performance by skipping structural comparisons. IMO this should not be summarily closed simply because it hasn't had attention recently.

This is the 9th most voted on PR in the entire repo, I  think that's a clear indicator that a lot of people want it and it's probably worth pushing forward. Would you please reopen it?

@weswigham since you authored this, do you think it is worth continuing the work you started or should this be closed in favor of a new discussion/implementation?

**Comment by weswigham** [comment]
I mean, personally, I think the structural version of this PR, the `tag` types PR, is the more general and better one. You can make those tags into nominal ones by including a unique symbol in your tag type, but conversely you can't make these into a structural tag, and structural tags have a lot of nice properties for interface merging and cross-library tag type management.

**Comment by evelant** [comment]
@weswigham I'm not familiar with the PR you're referring to, would you link to it please?

**Comment by weswigham** [comment]
#33290

**Comment by evelant** [comment]
Thanks, I agree that PR seems like it is a better approach. Unfortunately it looks like that one got summarily closed as well. Given that there seems to be a lot of interest in support for some form of nominal typing perhaps it should be reopened?

**Comment by sandersn** [comment]
From the last comment on the other PR:

> This is effectively an implementation sketch for https://github.com/microsoft/TypeScript/issues/202. We'd prefer people interact with the suggestion issues than the PRs, since the implementation of a feature is mechanical once the design has been worked out, and nominal types are still very much under discussion

**Comment by ruojianll** [comment]
@weswigham Hey great man, how was this going?
